cocos/core/pipeline/pipeline-state-manager.ts

```diff
@@ -0,0 +1,49 @@
+/**
+ * @hidden
+ */
+
+import { GFXPipelineState, IGFXPipelineStateInfo } from '../gfx/pipeline-state';
+import { IPSOCreationInfo } from '../renderer/scene/submodel';
+import { GFXRenderPass } from '../gfx/render-pass';
+import { GFXInputAssembler } from '../gfx/input-assembler';
+import { murmurhash2_32_gc } from '../utils/murmurhash2_gc';
+import { GFXDevice } from '../gfx/device';
+
+export class PipelineStateManager {
+    private static _PSOHashMap: Map<number, GFXPipelineState> = new Map<number, GFXPipelineState>();
+
+    static getOrCreatePipelineState(
+        device: GFXDevice,
+        psoCreateInfo: IPSOCreationInfo,
+        renderPass: GFXRenderPass,
+        ia: GFXInputAssembler
+        ): GFXPipelineState {
+
+            const hash1 = psoCreateInfo.hash;
+            const hash2= renderPass.hash;
```

Comment:
space

Reply:
Good catch, i don't know why VS code make the space like this.